hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/join/JoinPredicateScanResultSetIterator.java

```diff
@@ -0,0 +1,129 @@
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * This class adapts the internally fetched ResultSet and the specified rowMapper into an iterator
+ * The rowMapper has the SQL Join logic. As long as it returns an item, we can traverse this iterator
+ */
+public class JoinPredicateScanResultSetIterator<T> implements Iterator<T> {
+    private static final ILogger LOGGER = Logger.getLogger(JoinPredicateScanResultSetIterator.class);
+    private Connection connection;
+    private final String sql;
+    private final Function<ResultSet, T> rowMapper;
+    private final Consumer<PreparedStatement> preparedStatementSetter;
+    private boolean hasNext;
+    private boolean iteratorClosed;
+    private ResultSet resultSet;
+    private PreparedStatement preparedStatement;
+    private T nextItem;
+
+    public JoinPredicateScanResultSetIterator(Connection connection,
+                                              String sql,
+                                              Function<ResultSet, T> rowMapper,
+                                              Consumer<PreparedStatement> preparedStatementSetter) {
+        this.connection = connection;
+        this.sql = sql;
+        this.rowMapper = rowMapper;
+        this.preparedStatementSetter = preparedStatementSetter;
+    }
+
+    @Override
+    public boolean hasNext() {
+        try {
+            if (iteratorClosed) {
+                return false;
+            }
+            getNextItem();
+            if (!hasNext) {
+                close();
+            }
+            return hasNext;
+        } catch (SQLException sqlException) {
+            close();
+            throw ExceptionUtil.sneakyThrow(sqlException);
+        }
+    }
+
+    @Override
+    public T next() {
+        try {
+            if (iteratorClosed) {
+                throw new NoSuchElementException();
+            }
+            if (nextItem == null) {
+                getNextItem();
+            }
+            if (!hasNext) {
+                close();
+                throw new NoSuchElementException();
+            }
+            T result = nextItem;
+            nextItem = null;
+            return result;
+        } catch (SQLException sqlException) {
+            close();
+            throw ExceptionUtil.sneakyThrow(sqlException);
+        }
+    }
+
+    private void lazyInit() throws SQLException {
+        if (preparedStatement == null) {
+            preparedStatement = connection.prepareStatement(sql);
+            preparedStatementSetter.accept(preparedStatement);
+            resultSet = preparedStatement.executeQuery();
+        }
+    }
+
+    private void close() {
+        LOGGER.info("Close is called");
+        iteratorClosed = true;
+
+        IOUtil.closeResource(resultSet);
+        resultSet = null;
```

Comment:
Is nulling necessary here? The referenced objects still will be cleaned by GC, because they will not be accessible anymore.

Reply:
Normally no it is not necessary. But just in case there is an  bug in the code and close() is called again, there is a defense. The objects are null

Reply:
How do you think: would some null-check condition before resource closing (let's say, ` connection == null`) be better, or you like a current solution?

Reply:
IOUtils was checking for null. But since we added a new flag "iteratorClosed" maybe we don't need to assign fields to null
I deleted the assignments to null 
rev : 92b8d23155c211872167387fd532e313878e08d6